README.md

```diff
@@ -4,45 +4,84 @@ Data Engineering encompasses a broad range of disciplines aimed at facilitating
 
 
 # Data Extraction
-- Do incremental pipelines where applicable (depends on incremental extraction)
+- Do incremental pipelines where applicable (depends on incremental extraction).
 - Use generators for extraction to avoid filling memory. Manage the max memory size used by writing to disk buffers, ensuring scalability of your pipeline.
 - Easy local testing - your pipeline, or at least the extractor, should run locally without having to spin up anything heavy.
 - Handle errors produced by the source - be mindful of the sources' way of reporting errors - they may be silent.
 - Consider implementing "checkpoints" or chunking large extraction jobs to avoid crashes due to network issues.
 - Retry network issues or temporary server errors with incremental backoffs.
 - Apply data contracts if ingesting data from sources that might send trash (web events for example)
-
+- Make sure to extract data in compliance with privacy laws and maintain security and user consent.
+- Respect API rate limits when extracting data from external sources. Implementing rate-limiting strategies prevents overloading source systems and ensures a stable extraction process.
+- Document the extraction process and maintain metadata about the data source, extraction time, data volume, and any transformations applied. This aids in traceability and enables you to send lateness alerts for jobs that don't succeed within average times" always good to give specific examples.
+- Support flexible data formatting options during extraction to accommodate the needs of downstream processes.
+- Capture metadata in rich formats like Paruqet or Avro on extraction when possible, to enable downstream processes to use it.
 
 # Data Normalisation
 - Self deploying ddls and automated migrations: Infer your ddls from data or define them and keep them together with the pipeline; A pipeline that self deploys the needed ddl doesn't run the risk of the loading code being out of sync with db state.
 - Track schema changes and notify them to the producer and consumer, so the consumer can curate and the producer can react in case of accident.
 - Type your data as soon as you can - working with schemaless data is known as "schema on read" and introduces vulnerable assumptions into your code.
 - Automate your data typing where sensible - do not waste time guessing and manually maintaining those guesses.
```

Comment:
What is data typing? Is it data type of columns ??
Rewrite sentence:
Automate your data types where appropriate - avoid wasting time on speculation and manual upkeep of those assumptions.